velox/benchmarks/tpch/TpchGen.cpp

```diff
@@ -0,0 +1,164 @@
+
+    if (offset == 0) {
+      std::cout << std::endl
+                << "> first 10 rows from TPC-H " << tableName
+                << " table: " << data->toString() << std::endl;
+      std::cout << data->toString(0, 10) << std::endl;
+    }
+
+    writer.write(data);
+    writer.flush();
+    std::cout << "written offset " << offset << " (/" << numRows << ")"
+              << std::endl;
+  }
+
+  writer.close();
+}
+
+void generateOrdersAndLineitems(const std::filesystem::path& dataDirectory) {
+  auto pool = memory::getDefaultMemoryPool();
+  const std::string lineitemTableName =
+      std::string{tpch::toTableName(tpch::Table::TBL_LINEITEM)};
+  const std::string ordersTableName =
+      std::string{tpch::toTableName(tpch::Table::TBL_ORDERS)};
+
+  const std::filesystem::path& lineitemDirectory =
+      dataDirectory / lineitemTableName;
+  const std::filesystem::path& ordersDirectory =
+      dataDirectory / ordersTableName;
+  std::filesystem::create_directories(lineitemDirectory);
+  std::filesystem::create_directories(ordersDirectory);
+  const std::filesystem::path& lineitemFilePath =
+      lineitemDirectory / "001.parquet";
+  const std::filesystem::path& ordersFilePath = ordersDirectory / "001.parquet";
+
+  Writer lineitemWriter{
+      dwio::common::DataSink::create(lineitemFilePath), *pool, kRowsPerSplit};
+  Writer ordersWriter{
+      dwio::common::DataSink::create(ordersFilePath), *pool, kRowsPerSplit};
+
+  const size_t numOrderRows =
+      tpch::getRowCount(facebook::velox::tpch::Table::TBL_ORDERS, kScaleFactor);
+
+  size_t expectedLineitemRows = tpch::getRowCount(
+      facebook::velox::tpch::Table::TBL_LINEITEM, kScaleFactor);
+  size_t numGeneratedLineitemRows = 0;
+
+  // Make sure we include all rows and don't forget some at the end.
+  const size_t endOffset = numOrderRows + kRowsPerSplit;
+  for (size_t offset = 0; offset < endOffset; offset += kRowsPerSplit) {
+    auto orderData = getTpchData(
+        facebook::velox::tpch::Table::TBL_ORDERS,
+        kRowsPerSplit,
+        offset,
+        kScaleFactor,
+        pool.get());
+
+    auto lineitemData = getTpchData(
+        facebook::velox::tpch::Table::TBL_LINEITEM,
+        kRowsPerSplit,
+        offset,
+        kScaleFactor,
+        pool.get());
+
+    ordersWriter.write(orderData);
+    ordersWriter.flush();
+
+    numGeneratedLineitemRows += lineitemData->size();
+
+    lineitemWriter.write(lineitemData);
+    lineitemWriter.flush();
+
+    std::cout << "written offset " << offset << " (/" << numOrderRows << ")"
+              << std::endl;
+  }
+
+  VELOX_CHECK_EQ(numGeneratedLineitemRows, expectedLineitemRows);
+
+  ordersWriter.close();
+  lineitemWriter.close();
+}
+
+} // namespace
+
+int main() {
+  const std::filesystem::path dataDirectory = "/tmp/tpch-data ";
+  std::filesystem::create_directories(dataDirectory);
+
+  // We need to create these together, as lineitems are created per order.
```

Comment:
This is handled by then generator
https://github.com/facebookincubator/velox/blob/main/velox/tpch/gen/TpchGen.cpp#L412

Reply:
Maybe I misunderstand the method and its parameters, but the line in your link explicitly says `ordersOffset` even though we are generating the `lineitem` table. I think this is the reason why the code is not generating all 6 million rows, as the `lineitemOffset` is `> orderOffset` at some point (~ 1.5 million), at which point it does not generate any more data because we've reached the end of the SF.

Reply:
You are right. Lineitems requires the `ordersOffset` which is different from the `rowCount`. I fixed this and see the right count for SF 0.01. This should fix the SF 1 case as well.